### PR TITLE
Fixing Expansion Issue, Improving Documentation

### PR DESCRIPTION
**Pull request by StartAutomating**
Using Expanded Value
Improving Documentation
Retargeting to TemplateObject (because this test is valid in nested templates as well)

### CODE REVIEW

File: test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1 [hunk 1]
```diff
@@ -1,16 +1,26 @@
...
+param(
+# The template object (the contents of azureDeploy.json, converted from JSON)
 [Parameter(Mandatory=$true,Position=0)]
-$MainTemplateObject
-)
+$TemplateObject)
 
 # First, find all objects with an ID property in the MainTemplate.
-$ids = $MainTemplateObject  | Find-JsonContent -Key id -Value * -Like 
+$ids = $TemplateObject  | Find-JsonContent -Key id -Value * -Like
```
Comment (bmoore-msft) [inline test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: this should be for any property that endswith "id", e.g.
fooId
barId
id
etc

File: test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1 [hunk 2]
```diff
@@ -1,16 +1,26 @@
...
 
 foreach ($id in $ids) { # Then loop over each object with an ID
     $myId = "$($id.id)".Trim() # Grab the actual ID,
-    $expandedId = Expand-AzureRMTemplate -Expression $myId -InputObject $MainTemplateObject
-    if ($myId -notmatch '\[resourceId\(') { # check that it uses the ResourceID function
+    $expandedId = Expand-AzureRMTemplate -Expression $myId -InputObject $TemplateObject # then expand it.
+    if ($expandedId -notmatch '\[resourceId\(') { # Check that it uses the ResourceID function
```
Comment (bmoore-msft) [inline test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: we also need to allow for 
[variables('
[parameters('

we may have to flip it to a blacklist instead, but for now, allow those three things at leat...

Comment (StartAutomating) [reply test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: Actually we've already got this.  That's what the Expand-AzureRMTemplate is bringing to the table.

If it is a "resolvable collection", e.g. a top-level property, Expand-AzureRMTemplate will expand out the expression.

File: test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1 (hunk 1 above)
Comment (StartAutomating) [reply test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: Agreed.

I'll need to extent Find-JsonContent to do that and am trying to keep changes more discreet.  
Approve this one and I'll make an additional with the changes for Find-JsonContent.

Comment (bmoore-msft) [reply test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: do we have a workitem to track that?

File: test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1 (hunk 2 above)
Comment (bmoore-msft) [reply test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1]: here's what I'm getting when I run on the 100-marketplace-sample

    Cannot convert value "IDs-Should-Be-Derived-From-ResourceIDs.test.ps1" to type "System.Management.Automation.CommandMetadata". Error: "At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:3 char:45                                  +     Ensures that all IDs use the resourceID() function.                                                                                                                +                                             ~                                                                                                                          An expression was expected after '('.                                                                                                                                                                                                                                                                                                             At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:5 char:45    +     Ensures that all IDs use the resourceID() function, or resolve to ...                                                                                              +                                             ~                                                                                                                          An expression was expected after '('.                                                                                                                                                                                                                                                                                                             At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:5 char:119   + ... r resolve to parameters or variables that use the ResourceID() functi ...                                                                                          +                                                                  ~                                                                                                     An expression was expected after '('."

**Pull request merged**